### PR TITLE
Dataframe queries: support sparsely encoded dictionary arrays

### PR DESCRIPTION
**Pull request by teh-cmc**
And also clean up the `datatype` situation: I didn't know what should be in charge of specifying all the wrappers in the final datatype. Now I do.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7361?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7361?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/7361)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: crates/store/re_chunk/src/util.rs [hunk 1]
```diff
@@ -66,6 +70,103 @@ pub fn arrays_to_list_array(
...
+            .filter_map(move |(index, array)| {
+                if Some(index) == cur_index {
+                    None
+                } else {
+                    cur_index = Some(index);
+                    Some(array)
+                }
+            })
+            .collect_vec()
+    };
```
Comment (teh-cmc) [inline crates/store/re_chunk/src/util.rs:104]: surely there has to be a builtin helper for this

File: crates/store/re_chunk/src/util.rs [hunk 2]
```diff
@@ -66,6 +70,103 @@ pub fn arrays_to_list_array(
...
+                        cur_key += 1;
+                        cur_key
+                    } else {
+                        cur_index = Some(index);
+                        cur_key
+                    }
+                })
+            })
+            .collect_vec()
+    };
```
Comment (teh-cmc) [inline crates/store/re_chunk/src/util.rs:127]: ditto?

File: crates/store/re_chunk/src/util.rs [hunk 3]
```diff
@@ -66,6 +70,91 @@ pub fn arrays_to_list_array(
...
+            ArrowOffsets::try_from_lengths(arrays_dense_deduped.iter().map(|array| array.len()))
+                .unwrap();
+
+        ArrowListArray::<i32>::new(datatype, offsets.into(), data, None)
+    };
+
+    let datatype = ArrowDatatype::Dictionary(
+        arrow2::datatypes::IntegerType::UInt32,
+        std::sync::Arc::new(data.data_type().clone()),
+        false, // is_sorted
```
Comment (teh-cmc) [inline crates/store/re_chunk/src/util.rs:145]: yes it is

**Review by jleibs — APPROVED** [review]
Looks reasonable given the approach (tabled for a separate discussion/review)

As noted, I appreciate the fact that I believe our dictionary actually lets us differentiate:
- "repeated value in the data" => Different keys pointed at repeated values in the dictionary.
- from "repeated value via latest-it" => Should always have the same key.

**Pull request merged**